core/src/main/java/com/yahoo/oak/Chunk.java

```diff
@@ -717,67 +717,139 @@ void release() {
+    // The upper bound given by parameter key "to" might not be in this chunk
+    // thisChunkUpperBoundKey - is the minimal key of the next chunk,
+    // all current and future keys of this chunk are less than thisChunkUpperBoundKey
+    AscendingIter ascendingIter(ThreadContext ctx, K from, boolean fromInclusive, K to,
+        boolean toInclusive, OakScopedReadBuffer thisChunkUpperBoundKey) {
+        return new AscendingIter(ctx, from, fromInclusive, to, toInclusive, thisChunkUpperBoundKey);
     }
 
-    DescendingIter descendingIter(ThreadContext ctx) {
-        return new DescendingIter(ctx);
+    // Descending iterator from the end of the chunk.
+    // The lower bound given by parameter key "to" might not be in this chunk
+    DescendingIter descendingIter(ThreadContext ctx, K to, boolean toInclusive) {
+        return new DescendingIter(ctx, to, toInclusive);
     }
 
-    DescendingIter descendingIter(ThreadContext ctx, K from, boolean inclusive) {
-        return new DescendingIter(ctx, from, inclusive);
+    // Descending iterator from key "from" given as parameter.
+    // The lower bound given by parameter key "to" might not be in this chunk
+    DescendingIter descendingIter(ThreadContext ctx, K from, boolean fromInclusive, K to,
+        boolean toInclusive) {
+        return new DescendingIter(ctx, from, fromInclusive, to, toInclusive);
     }
 
-    private int advanceNextIndex(final int entryIndex) {
-        int next = entryIndex;
-        while (next != NONE_NEXT && !entrySet.isValueRefValidAndNotDeleted(next)) {
-            next = entrySet.getNextEntryIndex(next);
-        }
-        return next;
-    }
+    /********************************************************************************************/
+    /*------ Base Class for Chunk Iterators (keeps all the common fields and methods) ----------*/
 
-    interface ChunkIter {
-        boolean hasNext();
+    abstract class ChunkIter {
+        protected int next;         // index of the next entry to be returned
+        protected K to;             // stop bound key, or null if no stop bound
+        protected boolean toInclusive;  // inclusion flag for "to"
 
-        int next(ThreadContext ctx);
-    }
+        protected boolean needBoundCheckStatic = false; // to check the stop boundaries somewhere in this chunk?
+        protected boolean needBoundCheckDynamic = false; // to check the boundaries on each advancing?
+        protected int midIdx = sortedCount.get()/2; // approximately index of the middle key in the chunk
+
+        abstract boolean hasNext();
 
-    class AscendingIter implements ChunkIter {
+        abstract int next(ThreadContext ctx);
 
-        private int next;
+        abstract boolean isBoundCheckNeeded();
 
-        AscendingIter() {
+        abstract boolean afterStop(OakScopedReadBuffer boundKey);
+
+        protected void setStopCheckThreshold(
+            ThreadContext ctx, K to, boolean toInclusive, OakScopedReadBuffer boundKey) {
```

Comment:
I think that this brings the symmetry between ascending and descending scan. Descending scan has the internal (fast accessed) minKey that the scan can check upon. Otherwise for descending vs ascending scan we need different traversing logic that I would like to avoid. For now I do not apply all the suggestions related to this idea. We can discuss it during the code review meeting.